external/js/cothority/src/byzcoin/proof.ts

```diff
@@ -148,6 +189,10 @@ export default class Proof extends Message<Proof> {
             }
         }
 
+        // Update the known list of links. If we had access to the link-height, we could
+        // very much optimize this. But as it is, we can only store the current link list.
+        Proof.knownBlocks.set(genesisID, links.map((link) => link.to));
```

Comment:
Last point I'm just thinking of. If you ask for multiple proofs of the same byzcoin chain but very different blocks, you will kinda loose the benefit of what you did as you override the previous links. Like if a proof has a few links and the next one as a direct link from the genesis.

Also the array search is slow in Javascript compared to an object (map) access. Line 150.

All of that to propose that you convert the IDs into string and use it as a key:
```javascript
const register = knownBlocks.get(genesisID);
const id = buf2hex(link.to);
register[id] = null;

_.has(register, id) // lodash - true
_.has(register, "abc") // false
```

And you can update the register with new verified IDs. Just a thought, feel free to discard.

Reply:
> Last point I'm just thinking of. If you ask for multiple proofs of the same byzcoin chain but very different blocks, you will kinda loose the benefit of what you did as you override the previous links. Like if a proof has a few links and the next one as a direct link from the genesis.

If I suppose that the proofs I get will be either from the same block or from a later block (which sounds as a very reasonable assumption to me), there is not much I can do against this inefficiency.

In a next step, one could implement https://github.com/dedis/cothority/issues/1960 to make it better.

> Also the array search is slow in Javascript compared to an object (map) access. Line 150.
> 
> All of that to propose that you convert the IDs into string and use it as a key:
> 
> ```js
> const register = knownBlocks.get(genesisID);
> const id = buf2hex(link.to);
> register[id] = null;
> 
> _.has(register, id) // lodash - true
> _.has(register, "abc") // false
> ```
> 
> And you can update the register with new verified IDs. Just a thought, feel free to discard.

Hmm - two questions:
- is it slow even compared to the signature-verification?
- how would the map-definition look like with this new definition?

Reply:
It would have been a concern if we keep the previous block IDs but I guess your assumption is right so it doesn't matter.